### PR TITLE
Add `DeviceDerivatives` jacobian product calculator

### PR DESCRIPTION
**Pull request by albi3ro**
This PR follows on from #4435 , and adds a child class that can handle device-provided derivatives.

It has two private caches, `_results_cache` and `_jacs_cache`. This handles storing the jacobian between the forward pass and the backward pass. The caches will also store results and jacobians between successive calls to the same instance.

The class can handle both device interfaces, `qml.Device` and `qml.devices.Device`.

Name for the class is still up in the air.  This class takes the full jacobian from the device and then performs the dot product.  The class in #4637 takes the jacobian product itself from the device.

### CODE REVIEW

File: pennylane/_device.py [hunk 1]
```diff
@@ -561,6 +561,9 @@ def execute_and_gradients(self, circuits, method="jacobian", **kwargs):
             tuple[list[array[float]], list[array[float]]]: Tuple containing list of measured value(s)
             and list of Jacobians. Returned Jacobians should be of shape ``(output_shape, num_params)``.
         """
+        if self.tracker.active:
```
Comment (albi3ro) [inline pennylane/_device.py:564]: I order to test the caching, I need to know when derivatives are requested.

File: pennylane/interfaces/jacobian_products.py [hunk 2]
```diff
@@ -215,3 +229,312 @@ def compute_jacobian(self, tapes: Batch):
...
+                self._results_cache[tapes] = results
+
+            if tapes in self._jacs_cache:
+                if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
+                    logger.debug("Retrieving jacobian from cache.")
+                jacs = self._jacs_cache[tapes]
+            else:
+                raise NotImplementedError(
+                    "No path to cache results without caching jac. This branch should not occur."
+                )
```
Comment (dwierichs) [inline pennylane/interfaces/jacobian_products.py:441]: It took me a bit to understand why this is the case. Could add a comment
```suggestion
                # Here the jac was not cached but the results were. This can not happen because results are never
                # cached alone (note that in the else clause above computing only results, jac must already be present)
                raise NotImplementedError(
                    "No path to cache results without caching jac. This branch should not occur."
                )
```

File: pennylane/interfaces/jacobian_products.py [hunk 3]
```diff
@@ -215,3 +229,312 @@ def compute_jacobian(self, tapes: Batch):
...
+
+        Returns:
+            TensorLike: the vector jacobian product.
+
+        Side Effects:
+            caches the newly computed jacobian if it wasn't already present in the cache.
+
+        **Examples:**
+
+        For an instance of :class:`~.JacobianProductCalculator` ``jpc``, we have:
```
Comment (dwierichs) [inline pennylane/interfaces/jacobian_products.py:466]: See above
```suggestion
        For an instance of :class:`~.DeviceJacobians` ``jpc``, we have:
```

File: pennylane/interfaces/jacobian_products.py [hunk 4]
```diff
@@ -215,3 +229,312 @@ def compute_jacobian(self, tapes: Batch):
...
+
+        Returns:
+            TensorLike: the full jacobian
+
+        Side Effects:
+            caches the newly computed jacobian if it wasn't already present in the cache.
+
+        **Examples:**
+
+        For an instance of :class:`~.JacobianProductCalculator` ``jpc``, we have:
```
Comment (dwierichs) [inline pennylane/interfaces/jacobian_products.py:518]:
```suggestion
        For an instance of :class:`~.DeviceJacobians` ``jpc``, we have:
```

File: tests/gradients/finite_diff/test_spsa_gradient_shot_vec.py [hunk 5]
```diff
@@ -30,6 +30,7 @@
 
 default_shot_vector = (1000, 2000, 3000)
 many_shots_shot_vector = tuple([100000] * 3)
+np.random.seed(42)
```
Comment (dwierichs) [inline tests/gradients/finite_diff/test_spsa_gradient_shot_vec.py:33]: Unrelated change? Fixing the seed on the file level might not be optimal with distributed testing?

File: tests/interfaces/test_jacobian_products.py [hunk 6]
```diff
@@ -62,6 +70,70 @@ def test_transform_jacobian_product_basics(self):
...
+
+        device = qml.device("default.qubit")
+        config = qml.devices.ExecutionConfig(gradient_method="adjoint")
+
+        jpc = DeviceJacobians(device, {}, config)
+
+        assert jpc._device is device
+        assert jpc._execution_config is config
+        assert jpc._gradient_kwargs == {}
+        assert jpc._is_new_device
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:84]: iirc there is a recommendation in pylint to make these tests explicitly check identity with `True` and `False` :thinking: 
```suggestion
        assert jpc._is_new_device is True
```

File: tests/interfaces/test_jacobian_products.py [hunk 7]
```diff
@@ -62,6 +70,70 @@ def test_transform_jacobian_product_basics(self):
...
+        old device interface."""
+
+        device = qml.devices.DefaultQubitLegacy(wires=5)
+        gradient_kwargs = {"method": "adjoint_jacobian"}
+
+        jpc = DeviceJacobians(device, gradient_kwargs)
+
+        assert jpc._device is device
+        assert jpc._gradient_kwargs == gradient_kwargs
+        assert not jpc._is_new_device
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:101]:
```suggestion
        assert jpc._is_new_device is False
```

File: tests/interfaces/test_jacobian_products.py [hunk 8]
```diff
@@ -62,6 +70,70 @@ def test_transform_jacobian_product_basics(self):
...
+            r" ExecutionConfig(grad_on_execution=None, use_device_gradient=None,"
+            r" gradient_method='adjoint', gradient_keyword_arguments={},"
+            r" device_options={}, interface=None, derivative_order=1)>"
+        )
+
+        assert repr(jpc) == expected
+
+    @pytest.mark.parametrize("jpc", dev_jpc_matrix)
+    def test_no_shot_vector_with_dev_jacs(self, jpc):
+        """Test that device derivatives with shot vectors raise a not implemented error."""
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:125]:
```suggestion
        """Test that device derivatives with shot vectors raise a NotImplementedError."""
```

File: tests/interfaces/test_jacobian_products.py [hunk 9]
```diff
@@ -91,6 +164,207 @@ def test_jacobian_basic(self, jpc):
...
+
+        assert qml.math.allclose(jacs[0], -np.sin(phi))
+        assert qml.math.allclose(jacs[1][0][0], 0)
+        assert qml.math.allclose(jacs[1][0][1], np.cos(y))
+        assert qml.math.allclose(jacs[1][1][0], -np.cos(x) * np.sin(y))
+        assert qml.math.allclose(jacs[1][1][1], -np.sin(x) * np.cos(y))
+
+
+@pytest.mark.parametrize("jpc", dev_jpc_matrix)
+class TestCaching:
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:244]:
```suggestion
class TestCachingDeviceJacobians:
```

File: tests/interfaces/test_jacobian_products.py [hunk 10]
```diff
@@ -91,6 +164,207 @@ def test_jacobian_basic(self, jpc):
...
+        assert qml.math.allclose(results[0], np.cos(0.1))
+        assert jpc._device.tracker.totals["execute_and_derivative_batches"] == 1
+        assert jpc._device.tracker.totals["derivatives"] == 1
+
+        # extra execution since needs to do the forward pass again.
+        assert (
+            jpc._device.tracker.totals["executions"] == 2
+            if isinstance(jpc._device, qml.Device)
+            else 1
+        )
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:264]:
```suggestion
        expected_execs = 1 if jpc._is_new_device else 2
        assert jpc._device.tracker.totals["executions"] == expected_execs
```

File: pennylane/interfaces/jacobian_products.py [hunk 11]
```diff
@@ -215,3 +229,312 @@ def compute_jacobian(self, tapes: Batch):
...
+    >>> gradient_kwargs = {"method": "adjoint_jacobian"}
+    >>> jpc_lightning = DeviceJacobians(device, gradient_kwargs)
+
+    **Technical comments on caching and calculating the gradients on execution:**
+
+    In order to store results and jacobian for the backward pass during the forward pass,
+    the ``_jacs_cache`` and ``_results_cache`` properties are ``LRUCache`` objects with a maximum size of 10.
+
+    Note that since the hash and equality of :class:`~.QuantumScript` is based on object location, not contents, the
+    caching and retrieval should be more performant than a lookup based on the potentially expensive ``QuantumScript.hash``.
```
Comment (dwierichs) [inline pennylane/interfaces/jacobian_products.py:263]: I'm not sure I follow here. The loookup _is_ based on `QuantumScript.hash`, which is based on the object location (which is not expensive, is it?)? If not, what is it based on?

File: tests/interfaces/test_jacobian_products.py [hunk 12]
```diff
@@ -91,6 +164,207 @@ def test_jacobian_basic(self, jpc):
...
+        assert qml.math.allclose(jvps, -0.5 * np.sin(0.8))
+
+        assert jpc._results_cache[batch] is res
+        assert qml.math.allclose(jpc._jacs_cache[batch], (-np.sin(0.8)))
+
+        with jpc._device.tracker:
+            jpc.execute_and_compute_jvp(batch, tangents)
+
+        assert jpc._device.tracker.totals.get("derivatives", 0) == 0
+        assert jpc._device.tracker.totals.get("executions", 0) == 0
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:327]: Could add
```suggestion
        assert jpc._device.tracker.totals.get("executions", 0) == 0
        with jpc._device.tracker:
            jpc.execute(batch)
        assert jpc._device.tracker.totals.get("executions", 0) == 0
```

File: tests/interfaces/test_jacobian_products.py [hunk 13]
```diff
@@ -91,6 +164,207 @@ def test_jacobian_basic(self, jpc):
...
+        tape1 = qml.tape.QuantumScript([qml.RZ(0.5, wires=0)], [qml.expval(qml.PauliX(0))])
+        batch = (tape1,)
+        dy = ((0.5,),)
+
+        with jpc._device.tracker:
+            jpc.compute_vjp(batch, dy)
+
+        assert jpc._device.tracker.totals.get("executions", 0) == (
+            1 if isinstance(jpc._device, qml.Device) else 0
+        )
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:341]:
```suggestion
        assert jpc._device.tracker.totals.get("executions", 0) == int(jpc._is_new_device)
```

File: tests/interfaces/test_jacobian_products.py [hunk 14]
```diff
@@ -91,6 +164,207 @@ def test_jacobian_basic(self, jpc):
...
+
+        assert batch not in jpc._results_cache
+        assert qml.math.allclose(jpc._jacs_cache[batch], 0)
+
+        with jpc._device.tracker:
+            jpc.execute_and_compute_jvp(batch, ((0.5,),))
+
+        assert jpc._device.tracker.totals["executions"] == 1
+        assert jpc._device.tracker.totals.get("derivatives", 0) == 0
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:350]: Could add
```suggestion
        assert jpc._device.tracker.totals.get("derivatives", 0) == 0
        assert qml.math.allclose(jpc._results_cache[batch], 0)
```

File: tests/interfaces/test_jacobian_products.py [hunk 15]
```diff
@@ -91,6 +164,207 @@ def test_jacobian_basic(self, jpc):
...
+
+    def test_error_cant_cache_results_without_jac(self, jpc):
+        """Test that a not implemented error is raised if somehow the results are present
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:353]:
```suggestion
        """Test that a NotImplementedError is raised if somehow the results are cached
```

File: tests/interfaces/test_jacobian_products.py [hunk 16]
```diff
@@ -91,6 +164,207 @@ def test_jacobian_basic(self, jpc):
...
+        tape = qml.tape.QuantumScript([], [qml.state()])
+        batch = (tape,)
+        jpc._results_cache[batch] = "value"
+
+        with pytest.raises(NotImplementedError):
+            jpc.execute_and_compute_jvp(batch, tuple())
+
+
+@pytest.mark.parametrize("jpc", transform_jpc_matrix)
+class TestProbsOut:
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:365]:
```suggestion
class TestProbsTransformJacobians:
```

File: tests/test_qubit_device.py [hunk 17]
```diff
@@ -1426,7 +1426,7 @@ def test_tracker_multi_execution(self, dev_name):
         ):
             assert tracked_r == expected_r
 
-    @pytest.mark.all_interfaces
+    @pytest.mark.autograd
```
Comment (dwierichs) [inline tests/test_qubit_device.py:1429]: Is this just to save resources?

File: pennylane/interfaces/jacobian_products.py [hunk 18]
```diff
@@ -215,3 +229,312 @@ def compute_jacobian(self, tapes: Batch):
...
+                device,
+                execution_config,
+                gradient_kwargs,
+            )
+
+        self._device = device
+        self._execution_config = execution_config
+        self._gradient_kwargs = gradient_kwargs
+
+        self._is_new_device = not isinstance(device, qml.Device)
```
Comment (dwierichs) [inline pennylane/interfaces/jacobian_products.py:318]: Not a major thing but while reviewing the tests I noticed that `jpc._is_new_device` reads a bit funny if `jpc` is not a device. Maybe 
```suggestion
        self._uses_new_device = not isinstance(device, qml.Device)
```

**Review by dwierichs — COMMENTED** [review]
Thanks @albi3ro, very nice addition of the device derivatives! I really like the modular structure of this new interface, which also shines in the test suite: Very clean and readable. :100: 

I had a number of comments about docs/code comments, and about the naming of the class and the `_is_new_device` attribute.

File: doc/releases/changelog-dev.md [hunk 19]
```diff
@@ -77,9 +77,10 @@
 
 <h3>Improvements 🛠</h3>
 
-* The `JacobianProductCalculator` abstract base class and implementation `TransformJacobianProducts`
-  have been added to `pennylane.interfaces.jacobian_products`.
+* The `JacobianProductCalculator` abstract base class and implementations `TransformJacobianProducts`
+  and `DeviceDerivatives` have been added to `pennylane.interfaces.jacobian_products`.
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply doc/releases/changelog-dev.md:81]: I see that the Calculator introduced in #4637 carries the corresponding name. 
I think we should discuss the naming here, because it is currently based on the underlying computational strategy rather than the computed quantities, which may lead to confusion imho.
Alternatively, we could be explicit in the differences between the two added classes.

[the remark this answers is not in the document]
Comment (albi3ro) [reply doc/releases/changelog-dev.md:81]: As you pointed out, we will end up with two classes:

1. Uses device provided jacobians and then performs the product (this PR)
2. Uses device provided jvp and vjps. (next pr)

I'd be happy to get suggestions for the names.  Maybe `DeviceDerivatives` for this one? Then its more of a break from the "jacobian product" name?

File: pennylane/interfaces/jacobian_products.py (hunk 11 above)
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:263]: I'm trying to specify that we are *not* using `QuantumScript.hash`, as that is expensive and not worth it. So you could have two tapes that are "identical", but have different locations in memory and will be separate caches.

File: pennylane/interfaces/jacobian_products.py [hunk 20]
```diff
@@ -215,3 +229,312 @@ def compute_jacobian(self, tapes: Batch):
...
+
+    **Technical comments on caching and calculating the gradients on execution:**
+
+    In order to store results and jacobian for the backward pass during the forward pass,
+    the ``_jacs_cache`` and ``_results_cache`` properties are ``LRUCache`` objects with a maximum size of 10.
+
+    Note that since the hash and equality of :class:`~.QuantumScript` is based on object location, not contents, the
+    caching and retrieval should be more performant than a lookup based on the potentially expensive ``QuantumScript.hash``.
+    This means that the batch of tapes must be the
+    same instance, not just something that looks the same but has a different location in memory.
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:265]: Originally I had been using `id(batch)`, but that was showing flaky problems on the CI. I originally wanted to keep this independent of our choice of `__eq__` and `__hash__` so we could change them in the future. I'll go back and update the documentation to try and make it clearer.

File: pennylane/interfaces/jacobian_products.py [hunk 21]
```diff
@@ -215,3 +229,312 @@ def compute_jacobian(self, tapes: Batch):
...
+
+        Returns:
+            ResultBatch, TensorLike: the results of the execution and the jacobian vector product
+
+        Side Effects:
+            caches newly computed results or jacobians if they were not already cached.
+
+        **Examples:**
+
+        For an instance of :class:`~.JacobianProductCalculator` ``jpc``, we have:
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:396]: Mostly a pure copy from the original docstring with the addition of the `Side Effect` information.

File: tests/gradients/finite_diff/test_spsa_gradient_shot_vec.py (hunk 5 above)
Comment (albi3ro) [reply tests/gradients/finite_diff/test_spsa_gradient_shot_vec.py:33]:
```suggestion
```

File: tests/test_qubit_device.py (hunk 17 above)
Comment (albi3ro) [reply tests/test_qubit_device.py:1429]: think just something I noticed while going through the tests.  This is an autograd test, not a all interfaces test.

File: pennylane/interfaces/jacobian_products.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/interfaces/jacobian_products.py]: Thanks  :)

File: pennylane/interfaces/jacobian_products.py [hunk 22]
```diff
@@ -215,3 +229,318 @@ def compute_jacobian(self, tapes: Batch):
...
+
+    Then when the vjp, jvp, or jacobian is requested, that cached value is used instead of requesting from
+    the device again.
+
+    >>> with device.tracker:
+    ...     vjp = jpc.compute_vjp(batch , (0.5, ) )
+    >>> vjp
+    (array([-0.42073549]),)
+    >>> device.tracker.totals
+    {}
```
Comment (josh146) [inline pennylane/interfaces/jacobian_products.py:290]: @albi3ro for a user working with a QNode, where `diff_method="adjoint"`, will the cache be reset between each execution?

Or will requesting the same derivatives across multiple executions result in the cache being retained and used?

I believe the current implementation, the cache is contained within the execution 'closure', so is only active per individual (forward+backward) pass

File: pennylane/interfaces/jacobian_products.py [hunk 23]
```diff
@@ -215,3 +229,318 @@ def compute_jacobian(self, tapes: Batch):
...
+
+    **Lack of shot vector support:**
+
+    Given we currently do not have any examples of device derivatives with shots vectors (or even finite shots),
+    we are bypassing support for partitioned shots.
```
Comment (josh146) [inline pennylane/interfaces/jacobian_products.py:295]: @albi3ro what do you mean by this?

> (or even finite shots)

Comment (josh146) [reply pennylane/interfaces/jacobian_products.py:295]: Is this because the adjoint method for lightning only supports `shots=None`?

File: pennylane/interfaces/jacobian_products.py (hunk 22 above)
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:290]: A new instance is created on each call to `qml.execute`.

File: pennylane/interfaces/jacobian_products.py (hunk 23 above)
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:295]: adjoint on both default qubit and lightning are only supported for `shots=None`.

File: pennylane/interfaces/jacobian_products.py (hunk 22 above)
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:290]: I could potentially add a `cache` keyword argument and allow it to be reused between computations.

Comment (josh146) [reply pennylane/interfaces/jacobian_products.py:290]: > A new instance is created on each call to qml.execute.

Got it! 

> I could potentially add a cache keyword argument and allow it to be reused between computations.

one to double check with @trbromley beforehand, because it could be that there is no user need for this right now :) I was just more curious if this was a change in existing behaviour or not.

File: pennylane/interfaces/jacobian_products.py (hunk 23 above)
Comment (josh146) [reply pennylane/interfaces/jacobian_products.py:295]: I wonder if it could be tested with minimal effort through mocking? E.g., create a mock device that does nothing but return an arbitrary ndarray for finite shot jacobian for any circuit.

Or, mock default.qubit to return an arbitrary ndarray if the finite shot device jacobian is requested.

Comment (josh146) [reply pennylane/interfaces/jacobian_products.py:295]: E.g., don't test that the result is correct, just check that the pipeline works.

File: pennylane/interfaces/jacobian_products.py (hunk 22 above)
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:290]: Basically this is the way to save jacobians for the backward pass.

File: pennylane/interfaces/jacobian_products.py (hunk 23 above)
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py]: I've added a `ParamShiftDerivativesDevice` to `tests/param_shift_device_dev.py`.  This device uses parameter shift to product "device" derivatives. With this, we can test device derivatives and jacobian products with shot vectors.

I was able to add it for the vjp, since I had some reference to work with in `autograd.py`, but I'm still trying to figure out how to integrate jvp's with shot vectors.

**Review by dwierichs — COMMENTED** [review]
Nice future-facing addition to include shot-based device gradients. Maybe we'll come across an application for this! Or maybe users will have an application where they want to provide something like the `ParameterShiftDerivativesDevice` :)
Made a suggestion for the JVP part, hope that it works!

File: pennylane/interfaces/jacobian_products.py (hunk 22 above)
Comment (trbromley) [reply pennylane/interfaces/jacobian_products.py:306]: My preference is to preserve existing behaviour (as closely as possible), as I don't recall hearing anything from users asking for changes.

File: pennylane/interfaces/jacobian_products.py [hunk 24]
```diff
@@ -30,6 +32,34 @@
...
+
+        vjps.append(qml.math.sum(qml.math.stack(shot_vjps), axis=0))
+
+    return tuple(vjps)
+
+
+def _compute_jvps(jacs, tangents, multi_measurements, has_partitioned_shots):
+    """Compute the jvps of multiple tapes, directly for a Jacobian and tangents."""
+    if has_partitioned_shots:
+        raise NotImplementedError("not sure how to make this work yet.")
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:58]: Modified it a little, but seems to be working!

File: tests/interfaces/test_jacobian_products.py [hunk 25]
```diff
@@ -62,34 +86,369 @@ def test_transform_jacobian_product_basics(self):
...
+        tape = qml.tape.QuantumScript([qml.RX(x, 0)], [qml.expval(qml.PauliZ(0))], shots=shots)
         tangents = ((0.5,),)
         res, jvp = jpc.execute_and_compute_jvp((tape,), tangents)
-        assert qml.math.allclose(res[0], np.cos(x))
-        assert qml.math.allclose(jvp[0], -0.5 * np.sin(x))
 
-    def test_vjp_basic(self, jpc):
+        if tape.shots.has_partitioned_shots:
+            assert len(res[0]) == 2
+            print(jvp[0])
```
Comment (dwierichs) [inline tests/interfaces/test_jacobian_products.py:157]:
```suggestion
```

**Review by dwierichs — APPROVED** [review]
Just found a `print` statement, otherwise looks good to me, thanks @albi3ro ! :tada:

File: pennylane/interfaces/jacobian_products.py [hunk 26]
```diff
@@ -215,3 +242,310 @@ def compute_jacobian(self, tapes: Batch):
         )
         results = self._inner_execute(jac_tapes)
         return tuple(batch_post_processing(results))
+
+
+class DeviceDerivatives(JacobianProductCalculator):
```
Comment (rmoyard) [inline pennylane/interfaces/jacobian_products.py:247]: Not necessarily in this PR: but how do we choose between the methods in the interfaces? Let's say you can have all three available jacobian available, vjp with multiplication/without multiplication and jvp with multiplication/without multiplication?

File: pennylane/interfaces/jacobian_products.py [hunk 27]
```diff
@@ -215,3 +242,310 @@ def compute_jacobian(self, tapes: Batch):
...
+
+        """
+        if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
+            logger.debug("Forward pass called with %s", tapes)
+        results, jac = self._dev_execute_and_compute_derivatives(tapes)
+        self._results_cache[tapes] = results
+        self._jacs_cache[tapes] = jac
+        return results
+
+    def execute_and_compute_jvp(self, tapes: Batch, tangents):
```
Comment (rmoyard) [inline pennylane/interfaces/jacobian_products.py:390]: Are we not missing `execute_and_compute_vjp`?

Comment (rmoyard) [inline pennylane/interfaces/jacobian_products.py:390]: Are we not missing `execute_and_compute_vjp`?

**Review by rmoyard — COMMENTED** [review]
Thanks for this PR! Two high level questions, how do we integrate `adjoint_vjp` and `adjoint_jvp` that do not use the multiplication? Second question: how do we choose between the tmethods of this class in the interfaces when they are multiple choise like (adjoint_vjp and vjp with multiplication)?

File: pennylane/interfaces/jacobian_products.py [hunk 28]
```diff
@@ -215,3 +242,310 @@ def compute_jacobian(self, tapes: Batch):
...
+        multi_measurements = (len(t.measurements) > 1 for t in tapes)
+        jvps = _compute_jvps(
+            jacs,
+            tangents,
+            multi_measurements,
+            has_partitioned_shots=tapes[0].shots.has_partitioned_shots,
+        )
+        return results, jvps
+
+    def compute_vjp(self, tapes, dy):
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:464]: See the next PR: https://github.com/PennyLaneAI/pennylane/pull/4637

That `DeviceJacobianProducts` class will be used when we want to use device provided vjps instead of device provided jacobians.

File: pennylane/interfaces/jacobian_products.py (hunk 26 above)
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:247]: Autograd and torch will use `compute_vjp`. 

Tensorflow will sometimes use `compute_vjp` and sometimes use `compute_jacobian`, based on whether or not it wants to trace the calculation.

Jax will use `execute_and_compute_jvp` and `compute_vjp`.  

Jax-jit will use `compute_jacobian`.

File: pennylane/interfaces/jacobian_products.py (hunk 27 above)
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:390]: No interface will bind `execute_and_compute_vjp`, as you can't really have the `dy` until you perform the forward pass. If in the future we need it for a situation, then we can add that method in.

File: pennylane/interfaces/jacobian_products.py [hunk 29]
```diff
@@ -215,3 +242,310 @@ def compute_jacobian(self, tapes: Batch):
...
+        Converts tapes to numpy before computing the derivatives on the device.
+
+        Dispatches between the two different device interfaces.
+        """
+        numpy_tapes = tuple(qml.transforms.convert_to_numpy_parameters(t) for t in tapes)
+        if self._uses_new_device:
+            return self._device.compute_derivatives(numpy_tapes, self._execution_config)
+        return self._device.gradients(numpy_tapes, **self._gradient_kwargs)
+
+    def execute(self, tapes: Batch):
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:375]: Maybe `forward_pass` or `execute_and_cache_jacobian`?

File: pennylane/interfaces/jacobian_products.py [hunk 30]
```diff
@@ -215,3 +242,310 @@ def compute_jacobian(self, tapes: Batch):
...
+                raise NotImplementedError(
+                    "No path to cache results without caching jac. This branch should not occur."
+                )
+
+        multi_measurements = (len(t.measurements) > 1 for t in tapes)
+        jvps = _compute_jvps(
+            jacs,
+            tangents,
+            multi_measurements,
+            has_partitioned_shots=tapes[0].shots.has_partitioned_shots,
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py]: I originally just copied these functions from the interface files, but I just updated them to accept ragged partitioning.

**Comment by albi3ro** [comment]
> Thanks for this PR! Two high level questions, how do we integrate `adjoint_vjp` and `adjoint_jvp` that do not use the multiplication? Second question: how do we choose between the tmethods of this class in the interfaces when they are multiple choise like (adjoint_vjp and vjp with multiplication)?

Thanks for the review :)

`DefaultQubit.compute_vjp` and `DefaultQubit.execute_and_compute_jvp` will be bound through the next class to be added: https://github.com/PennyLaneAI/pennylane/pull/4637

Each interface will prefer a different method of the class. Autograd and torch will just use `compute_vjp`. Jax jit will use `compute_jacobian`, and tensorflow will either use `compute_vjp` or `compute_jacobian` based on whether or not it needs to trace the vjp calculation.

Jax will use `execute_and_compute_jvp` for all cases except for when its a `DeviceJacobianProducts` class (yes that's patchy but it works).

File: pennylane/interfaces/jacobian_products.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/interfaces/jacobian_products.py:375]: I'd personally prefer `execute_and_cache_jacobian` :)

File: pennylane/interfaces/jacobian_products.py (hunk 23 above)
Comment (trbromley) [reply pennylane/interfaces/jacobian_products.py]: > I'm still trying to figure out how to integrate jvp's with shot vectors

@albi3ro what was the resolution here?

File: pennylane/interfaces/jacobian_products.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/interfaces/jacobian_products.py:375]: Or just `cache_execute`.

[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/interfaces/jacobian_products.py:375]: Maybe long but  `execute_compute_derivatives_and_cache` is the most accurate, at the end I let @trbromley decide on that

File: pennylane/interfaces/jacobian_products.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/interfaces/jacobian_products.py:463]: Thank you I understand better :+1:

File: pennylane/interfaces/jacobian_products.py (hunk 26 above)
Comment (rmoyard) [reply pennylane/interfaces/jacobian_products.py:252]: This hypothetical but let's imagine a device developer comes with a device that can only compute the jacobians (like default qubit legacy for example), how do we make sure that it will work with our interfaces/execution. I think `compute_jacobian` should be the fallback for all devices. Also a user might want to choose between jacobians and vjp/jvps. Will there be a way to toggle between the methods?

File: pennylane/interfaces/jacobian_products.py (hunk 27 above)
Comment (rmoyard) [reply pennylane/interfaces/jacobian_products.py:395]: Thanks!

File: pennylane/interfaces/jacobian_products.py [hunk 31]
```diff
@@ -215,3 +247,303 @@ def compute_jacobian(self, tapes: Batch):
...
+
+    **Technical comments on caching and calculating the gradients on execution:**
+
+    In order to store results and Jacobians for the backward pass during the forward pass,
+    the ``_jacs_cache`` and ``_results_cache`` properties are ``LRUCache`` objects with a maximum size of 10.
+    In the current execution pipeline, only one batch will be used per instance, but a size of 10 adds some extra
+    flexibility for future uses.
+
+    Note that batches of identically looking :class:`~.QuantumScript` s that are different instances will be cached separately.
+    This is because the `hash` of  :class:`~.QuantumScript` is expensive, as it requires inspecting all its constituents,
```
Comment (mudit2812) [inline pennylane/interfaces/jacobian_products.py:285]:
```suggestion
    This is because the ``hash`` of  :class:`~.QuantumScript` is expensive, as it requires inspecting all its constituents,
```

File: pennylane/interfaces/jacobian_products.py (hunk 26 above)
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:252]: Yep.  See PR #4557: https://github.com/PennyLaneAI/pennylane/pull/4557

A device under the new interface will let pennylane know it prefers to use the device jacobian product by filling in `ExecutionConfig.use_device_jacobian_product = True`. 

https://github.com/PennyLaneAI/pennylane/blob/633c1c4cba801f408163361cec88a102a40881b7/pennylane/devices/qubit/preprocess.py#L415

When `qml.execute` sees that option is `True`, it will create a `DeviceJacobianProducts` instance.  Otherwise, it will create a `DeviceDerivatives` instance.

https://github.com/PennyLaneAI/pennylane/blob/633c1c4cba801f408163361cec88a102a40881b7/pennylane/interfaces/execution.py#L644

```
    if config.use_device_jacobian_product:
        jpc = DeviceJacobianProducts(device, config)

    elif config.use_device_gradient:
        jpc = DeviceDerivatives(device, config)
```

In the short term, I propose to add a keyword argument `use_device_jacobian_product` to first `qml.execute` and then `QNode`.  If `True`, then we will either use the device provided jacobian product or fail.  If `None`, we will query the device for what it wants to do and do that.  If `False`, we will use use the full jacobian.

https://github.com/PennyLaneAI/pennylane/blob/633c1c4cba801f408163361cec88a102a40881b7/pennylane/interfaces/execution.py#L414

We can start with the argument as `False`, and then switch it to `None` once we know the system is working well.


Lots of steps, but the guiding principles are:

* Interface code only pays attention to things that differ between interfaces.  `autograd.py` shouldn't care about the specifics of how pennylane computes the vjp.  It just should know how to bind a vjp to autograd.  This makes extending how we compute vjps easier.


* Devices should be able to configure what they deem "best".  By filling in things like `config.use_device_jacobian_product`, `config.use_device_gradient`, and `config.grad_on_execution`, a device tells pennylane what to do with it.


Hope that clears things up a little bit.

File: pennylane/interfaces/jacobian_products.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/interfaces/jacobian_products.py:375]: Since `cache_execute` has a pre-existing connotation with `qml.interfaces.cache_execute`, I'm going with `execute_and_cache_jacobian`.

File: pennylane/interfaces/jacobian_products.py (hunk 26 above)
Comment (rmoyard) [reply pennylane/interfaces/jacobian_products.py:252]: Thanks it looks good to me :+1:
